### PR TITLE
Export/import environments

### PR DESCRIPTION
**Pull request by atsikham**
Task #31

### CODE REVIEW

File: cmd/environments-export.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		if !isCurrentEnvUsed {
+			environments, err := environment.GetAll()
+			if err != nil {
+				logger.Fatal().Err(err).Msg("Unable to list all environments")
+			}
+			isEnvValid := false
+			for _, e := range environments {
+				if e.Uuid.String() == envID {
+					isEnvValid = true
+					debug("Found environment to export: %s", e.Uuid.String())
```
Comment (mkyc) [inline cmd/environments-export.go:75]: Please use direct `logger.Debug().Msgf(...)`.

**Review by mkyc — COMMENTED** [review]
So in general I would think about moving import/export logic to another package. Current version is good enough but it lacks additional logic: 
 - cleaning of directory before export
 - validation that this is correct environment directory before import

For now it's ok but if we would like to add this logic command file is not good place for it.

**Comment by atsikham** [comment]
> So in general I would think about moving import/export logic to another package. Current version is good enough but it lacks additional logic:
> 
> * cleaning of directory before export
> * validation that this is correct environment directory before import
> 
> For now it's ok but if we would like to add this logic command file is not good place for it.

Could you take a look again?

File: cmd/environments-export.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		if dstDir == "" {
+			path, err := os.Getwd()
+			if err != nil {
+				logger.Fatal().Err(err).Msg("Unable to get working directory")
+			}
+			dstDir = path
+		}
+
+		// Check if passed environment id is valid
+		if !isCurrentEnvUsed {
```
Comment (mkyc) [inline cmd/environments-export.go:63]: isn't it just else to `if` from line 44?

File: cmd/environments-import.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		}
+
+		srcFile = viper.GetString("from")
+
+	},
+
+	Run: func(cmd *cobra.Command, args []string) {
+		// Ask user for source file path if no file to export from is specified
+		if srcFile == "" {
+			srcFile, _ = promptui.PromptForString("File to export environment from")
```
Comment (mkyc) [inline cmd/environments-import.go:38]:
```suggestion
			srcFile, _ = promptui.PromptForString("File to import environment from")
```

File: cmd/environments-import.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			// Check if environment with such id is already in place
```
Comment (mkyc) [inline cmd/environments-import.go:39]:
```suggestion
			// Check if provided source file exists
```

File: cmd/environments-import.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			if _, err := os.Stat(srcFile); err != nil {
```
Comment (mkyc) [inline cmd/environments-import.go:40]: I think this `if` should be outside of first one. You want to check existence of file no matter how path was provided.

File: cmd/environments-import.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		if err != nil {
+			logger.Fatal().Err(err).Msg("Unable to import environment from specified file")
+		}
+
+		// Switch to the imported environment
+		config, err := configuration.GetConfig()
+		if err != nil {
+			logger.Fatal().Err(err).Msg("Get config failed")
+		}
+		err = config.SetUsedEnvironment(*envID)
```
Comment (mkyc) [inline cmd/environments-import.go:56]: I think you can add this `IsValid` method to be called inside `SetUsedEnvironment` if that is possible.

File: pkg/environment/environment.go [hunk 7]
```diff
@@ -287,3 +288,82 @@ func Get(uuid uuid.UUID) (*Environment, error) {
...
+	}
+	return isEnvValid, nil
+}
+
+//TODO add tests
+func Export(envID, dstDir string) error {
+	envPath := path.Join(util.UsedEnvironmentDirectory, envID)
+
+	// Final archive name is envID + .zip extension
+	err := archiver.Archive([]string{envPath}, path.Join(dstDir, envID)+".zip")
```
Comment (mkyc) [inline pkg/environment/environment.go:314]:
```suggestion
	err := archiver.Archive([]string{envPath}, path.Join(dstDir, envID+".zip"))
```
that feels more natural in my opinion

File: pkg/environment/environment.go [hunk 8]
```diff
@@ -287,3 +288,82 @@ func Get(uuid uuid.UUID) (*Environment, error) {
...
+			break
+		}
+	}
+	return isEnvValid, nil
+}
+
+//TODO add tests
+func Export(envID, dstDir string) error {
+	envPath := path.Join(util.UsedEnvironmentDirectory, envID)
+
```
Comment (mkyc) [inline pkg/environment/environment.go:312]: I think that somewhere here we should copy directory to temporary directory and then perform some `cleanDirectoryBeforeExport()` (this is dumb name) and package this cleaned directory.

File: pkg/environment/environment.go [hunk 9]
```diff
@@ -287,3 +288,82 @@ func Get(uuid uuid.UUID) (*Environment, error) {
...
+				return errors.New("Environment id is missing in the config")
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	envConfigDir := path.Join(util.GetHomeDirectory(), util.DefaultConfigurationDirectory, util.DefaultEnvironmentsSubdirectory)
```
Comment (mkyc) [inline pkg/environment/environment.go:347]: Isn't there already variable `util.UsedEnvironmentsSubdirectory` ... I believe there is.

**Review by mkyc — COMMENTED** [review]
Just some thoughts on current code stage

File: pkg/environment/environment.go (hunk 9 above)
Comment (atsikham) [reply pkg/environment/environment.go:347]: Yes, checked and there is `util.UsedEnvironmentDirectory`

**Review by mkyc — CHANGES_REQUESTED** [review]
Some little comments. I didn't read environment_test.go but I'll try to check it tomorrow in IDE directly form your branch.

File: cmd/cmd_test.go [hunk 10]
```diff
@@ -107,12 +107,13 @@ func TestCmd(t *testing.T) {
 			mockEnv:  true,
 			envId:    "cd7b59f8-6610-468a-8d56-3d1ea2566428",
 		},
-		{
+		// TODO: uncomment when environments removal is implemented
```
Comment (mkyc) [inline cmd/cmd_test.go:110]: why this is commented?

File: pkg/environment/environment_test.go [hunk 11]
```diff
@@ -527,53 +538,227 @@ func TestEnvironment_GetComponentByName(t *testing.T) {
...
-			if rc, ok := r.(string); ok {
-				if isWrongResult(t, errors.New(rc), wantErr) {
-					return
+func TestExport(t *testing.T) {
+	util.UsedConfigFile, util.UsedConfigurationDirectory, util.UsedEnvironmentDirectory = setup(t, "export")
+	util.UsedTempDirectory = path.Join(util.UsedConfigurationDirectory, util.DefaultEnvironmentsTempSubdirectory)
+	err := os.Mkdir(util.UsedTempDirectory, 0755)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (mkyc) [inline pkg/environment/environment_test.go:603]: isn't is better to add this to `setup()` method?

File: pkg/environment/environment_test.go [hunk 12]
```diff
@@ -527,53 +538,227 @@ func TestEnvironment_GetComponentByName(t *testing.T) {
...
+	err = importEnvInvalidSecond.Export(util.UsedConfigurationDirectory)
+	if err != nil {
+		t.Fatal(err)
+	}
+	importFileInvalidSecond := path.Join(util.UsedConfigurationDirectory, importEnvInvalidSecond.Uuid.String()+".zip")
+
+	// Remove environments to be able to export except one left by purpose
+	os.RemoveAll(path.Join(util.UsedEnvironmentDirectory, importEnvValidFirst.Uuid.String()))
+	os.RemoveAll(path.Join(util.UsedEnvironmentDirectory, importEnvValidSecond.Uuid.String()))
+	os.RemoveAll(path.Join(util.UsedEnvironmentDirectory, importEnvInvalidFirst.Uuid.String()))
```
Comment (mkyc) [inline pkg/environment/environment_test.go:722]: this whole code looks like it should go to separate function named i.e.: `prepareImportTestPrereqs()` or something like this.

File: pkg/environment/environment_test.go [hunk 13]
```diff
@@ -527,53 +538,227 @@ func TestEnvironment_GetComponentByName(t *testing.T) {
...
+	os.Remove(path.Join(util.UsedEnvironmentDirectory, importEnvInvalidFirst.Uuid.String(), util.DefaultEnvironmentConfigFileName))
+
+	importEnvInvalidSecond, err := Create("export-invalid2")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create archives
+	// Create a valid archive
+	err = importEnvValidFirst.Export(util.UsedConfigurationDirectory)
```
Comment (mkyc) [inline pkg/environment/environment_test.go:690]: This is optional suggestion, but I wouldn't do this here. I would run exactly code using packaging depandency here, and I would not use our method here. Reason is that if `Export` method will get broken you might have `Import` related test corrupted.

**Review by mkyc — CHANGES_REQUESTED** [review]
Answering your question in description of this PR, my opinion is that Import/Export is tested well enough in not to test it fully in `cmd/cmd_test.go`. I would test just invocation of method and incorrect invocation and check that outputs are as expected. So for example I would test `e environments export --id incorrect --destination /incorrect` and couple of similar, just to check error messages in communication with user.

File: cmd/cmd_test.go (hunk 10 above)
Comment (atsikham) [reply cmd/cmd_test.go:110]: It creates a new environment each tests execution and as a result you have it mixed up with yours environments.

File: cmd/environments-export.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+		var envID uuid.UUID
+
+		// Default environment and destination directory are current ones
+		// Check if environment is default
+		if envIDStr == "" {
+			config, err := configuration.GetConfig()
+			if err != nil {
+				logger.Fatal().Err(err).Msg("Unable to get environment config")
+			}
+			envID = config.CurrentEnvironment
```
[the remark this answers is not in the document]
Comment (atsikham) [reply cmd/environments-export.go:60]: great catch

File: cmd/environments-import.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+		envID, err := environment.Import(srcFile)
+		if err != nil {
+			logger.Fatal().Err(err).Msg("Unable to import environment from specified file")
+		}
+
+		// Switch to the imported environment
+		config, err := configuration.GetConfig()
+		if err != nil {
+			logger.Fatal().Err(err).Msg("Get config failed")
+		}
```
[the remark this answers is not in the document]
Comment (atsikham) [reply cmd/environments-import.go:56]: `Import()` returns a pointer, should that be changed?

[the remark this answers is not in the document]
Comment (mkyc) [reply cmd/environments-import.go:56]: I just meant to switch order maybe like following code: 

```golang
  		// Switch to the imported environment
 		config, err := configuration.GetConfig()
 		if err != nil {
 			logger.Fatal().Err(err).Msg("Get config failed")
 		}
		// Import environment
 		envID, err := environment.Import(srcFile)
 		if err != nil {
 			logger.Fatal().Err(err).Msg("Unable to import environment from specified file")
 		}
```

[the remark this answers is not in the document]
Comment (atsikham) [reply cmd/environments-import.go:56]: this was a comment in a wrong place, sorry.

File: pkg/configuration/configuration_test.go [hunk 16]
```diff
@@ -97,34 +102,44 @@ func TestConfig_SetUsedEnvironment(t *testing.T) {
...
 		t.Run(tt.name, func(t *testing.T) {
+			// Create necessary files and directories as environments are validated before switching
+			util.UsedConfigurationDirectory = tempDirectory
+			util.UsedEnvironmentDirectory = path.Join(util.UsedConfigurationDirectory, util.DefaultEnvironmentsSubdirectory)
+			envDirCurrent := path.Join(util.UsedEnvironmentDirectory, envIDCurrent)
+			envDirSwitch := path.Join(util.UsedEnvironmentDirectory, envIDSwitch)
+			os.MkdirAll(envDirCurrent, 0755)
+			os.Mkdir(envDirSwitch, 0755)
+			ioutil.WriteFile(path.Join(envDirCurrent, util.DefaultEnvironmentConfigFileName), []byte(fmt.Sprintf(envConfigTemplate, "env1", envIDCurrent)), 0644)
+			ioutil.WriteFile(path.Join(envDirSwitch, util.DefaultEnvironmentConfigFileName), []byte(fmt.Sprintf(envConfigTemplate, "env2", envIDSwitch)), 0644)
```
[the remark this answers is not in the document]
Comment (atsikham) [reply pkg/configuration/configuration_test.go:139]: Created a separate `prepareSetUsedEnvironmentPrereqs()` function as it's required only for this particular test.

**Review by mkyc — APPROVED** [review]
LGTM

**Review by seriva — APPROVED** [review]

**Review by mkyc — APPROVED** [review]

**Pull request merged**